pkg/ruler/evaluator_remote.go

```diff
@@ -161,36 +227,59 @@ func (q *remoteQuerier) query(ctx context.Context, query string, ts time.Time, l
+
+	if resp != nil {
+		instrument.ObserveWithExemplar(ctx, r.metrics.responseSizeBytes.WithLabelValues(orgID), float64(len(resp.Body)))
 	}
 
-	start := time.Now()
-	resp, err := q.client.Handle(ctx, &req)
+	log := log.With(logger, "query_hash", hash, "query", query, "instant", ts, "response_time", time.Since(start).String())
+
 	if err != nil {
-		level.Warn(logger).Log("msg", "failed to remotely evaluate query expression", "err", err, "query_hash", hash, "qs", query, "ts", ts, "response_time", time.Since(start).Seconds())
-		return nil, err
+		r.metrics.failedEvals.WithLabelValues("error", orgID).Inc()
+
+		level.Warn(log).Log("msg", "failed to evaluate rule", "err", err)
+		return nil, fmt.Errorf("rule evaluation failed: %w", err)
 	}
+
+	// TODO(dannyk): consider retrying if the rule has a very high interval, or the rule is very sensitive to missing samples
+	//   i.e. critical alerts or recording rules producing crucial RemoteEvaluatorMetrics series
 	if resp.Code/100 != 2 {
-		return nil, fmt.Errorf("unexpected response status code %d: %s", resp.Code, string(resp.Body))
+		r.metrics.failedEvals.WithLabelValues("upstream_error", orgID).Inc()
+
+		level.Warn(log).Log("msg", "rule evaluation failed with non-2xx response", "response_code", resp.Code, "response_body", resp.Body)
+		return nil, fmt.Errorf("unsuccessful/unexpected response - status code %d", resp.Code)
 	}
-	level.Debug(logger).Log("msg", "query expression successfully evaluated", "query_hash", hash, "qs", query, "ts", ts, "response_time", time.Since(start).Seconds())
 
-	return decodeResponse(resp)
+	maxSize := r.overrides.RulerRemoteEvaluationMaxResponseSize(orgID)
+	if maxSize > 0 && int64(len(resp.Body)) >= maxSize {
+		r.metrics.failedEvals.WithLabelValues("max_size", orgID).Inc()
+
+		level.Error(log).Log("msg", "rule evaluation exceeded max size", "max_size", maxSize, "response_size", len(resp.Body))
+		return nil, fmt.Errorf("%d bytes exceeds response size limit of %d (defined by ruler_remote_evaluation_max_response_size)", len(resp.Body), maxSize)
+	}
+
+	level.Debug(log).Log("msg", "rule evaluation succeeded")
+	r.metrics.successfulEvals.WithLabelValues(orgID).Inc()
+
+	return r.decodeResponse(ctx, resp, orgID)
 }
 
-func decodeResponse(resp *httpgrpc.HTTPResponse) (*logqlmodel.Result, error) {
+func (r *RemoteEvaluator) decodeResponse(ctx context.Context, resp *httpgrpc.HTTPResponse, orgID string) (*logqlmodel.Result, error) {
 	var decoded loghttp.QueryResponse
 	if err := json.NewDecoder(bytes.NewReader(resp.Body)).Decode(&decoded); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unexpected body encoding, not valid JSON: %w", err)
 	}
-	if decoded.Status == "error" {
-		return nil, fmt.Errorf("query response error: %s", decoded.Status)
+	if decoded.Status != loghttp.QueryStatusSuccess {
+		return nil, fmt.Errorf("query response error - status %q: %s", decoded.Status, resp.Body)
```

Comment:
resp.Body will potentially yield a very long unreadable byte slice. Maybe convert it to `string` first?

Reply:
It's being interpolated as a string in the `fmt.Errorf`.

Reply:
Good point. Also I believe `resp.Body` will be of type `io.ReadCloser`, so first reading it with something like `ioutil.ReadAll` and converting to string before printing in logs is the correct way to go I think.

even better to use `io.LimitReader` with some upper limit (just in case for some reason body is huge and to avoid some crazy memory footprints).